File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/reducers/LinearReducer.cpp
/*
 * LinearReducer.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include <iterator>
#include <list>
#include <iostream>
#include <algorithm>
#include "LinearReducer.h"
#include "Reducer.h"

LinearReducer::LinearReducer(std::vector<Point>& list_of_points)
: Reducer(list_of_points) {
}

void LinearReducer::reduce() {
  if (points.size() < 3) return;

  size_t write_idx = 1;
  for (size_t i = 2; i < points.size(); ++i) {
    const Point& start_p = points[write_idx - 1];
    const Point& end_p   = points[write_idx];
    const Point& curr_p  = points[i];

    int dx1 = end_p.x - start_p.x;
    int dy1 = end_p.y - start_p.y;

    int dx2 = curr_p.x - end_p.x;
    int dy2 = curr_p.y - end_p.y;

    int sx1 = (0 < dx1) - (dx1 < 0);
    int sy1 = (0 < dy1) - (dy1 < 0);

    int sx2 = (0 < dx2) - (dx2 < 0);
    int sy2 = (0 < dy2) - (dy2 < 0);

    if (sx1 == sx2 && sy1 == sy2) {
      points[write_idx] = curr_p;
    } else {
      write_idx++;
      points[write_idx] = curr_p;
    }
  }
  points.resize(write_idx + 1);
}

File: lib/contrek/reducers/linear_reducer.rb
module Contrek
  module Reducers
    class LinearReducer < Reducer
      def reduce!
        start_p = @points[0]
        end_p = @points[1]
        dir = seq_dir(start_p, end_p)
        @points[2..].map.with_index do |point, i|
          if (act_seq = seq_dir(end_p, point)) == dir
            @points.delete_at(@points.index(end_p))
          else
            dir = act_seq
          end
          end_p = point
        end
      end

      private

      def seq_dir(a, b)
        [b[:x] - a[:x], b[:y] - a[:y]]
      end
    end
  end
end
